server/src/test/java/org/elasticsearch/threadpool/ScalingThreadPoolTests.java

```diff
@@ -173,6 +182,125 @@ public void testScalingThreadPoolThreadsAreTerminatedAfterKeepAlive() throws Int
+            getTestName().toLowerCase(Locale.ROOT),
+            min,
+            max,
+            randomLongBetween(0, 100),
+            TimeUnit.MILLISECONDS,
+            rejectAfterShutdown,
+            EsExecutors.daemonThreadFactory(getTestName().toLowerCase(Locale.ROOT)),
+            new ThreadContext(Settings.EMPTY)
+        );
+        try {
+            final AtomicInteger executed = new AtomicInteger();
+            final AtomicInteger rejected = new AtomicInteger();
+            final AtomicInteger failed = new AtomicInteger();
+
+            final CountDownLatch latch = new CountDownLatch(max);
+            final CountDownLatch block = new CountDownLatch(1);
+            for (int i = 0; i < max; i++) {
+                execute(scalingExecutor, () -> {
+                    try {
+                        latch.countDown();
+                        block.await();
+                    } catch (InterruptedException e) {
+                        fail(e.toString());
+                    }
+                }, executed, rejected, failed);
+            }
+            latch.await();
+
+            assertThat(scalingExecutor.getCompletedTaskCount(), equalTo(0L));
+            assertThat(scalingExecutor.getActiveCount(), equalTo(max));
+            assertThat(scalingExecutor.getQueue().size(), equalTo(0));
+
+            final int queued = randomIntBetween(1, 100);
+            for (int i = 0; i < queued; i++) {
+                execute(scalingExecutor, () -> {}, executed, rejected, failed);
+            }
+
+            assertThat(scalingExecutor.getCompletedTaskCount(), equalTo(0L));
+            assertThat(scalingExecutor.getActiveCount(), equalTo(max));
+            assertThat(scalingExecutor.getQueue().size(), equalTo(queued));
+
+            scalingExecutor.shutdown();
+
+            final int queuedAfterShutdown = randomIntBetween(1, 100);
+            for (int i = 0; i < queuedAfterShutdown; i++) {
+                execute(scalingExecutor, () -> {}, executed, rejected, failed);
+            }
+            assertThat(scalingExecutor.getQueue().size(), rejectAfterShutdown ? equalTo(queued) : equalTo(queued + queuedAfterShutdown));
+
+            block.countDown();
+
+            assertBusy(() -> assertTrue(scalingExecutor.isTerminated()));
+            assertThat(scalingExecutor.getActiveCount(), equalTo(0));
+            assertThat(scalingExecutor.getQueue().size(), equalTo(0));
+            assertThat(
+                scalingExecutor.getCompletedTaskCount(),
+                rejectAfterShutdown ? equalTo((long) max + queued) : greaterThanOrEqualTo((long) max + queued)
+            );
+            assertThat(
+                ((EsRejectedExecutionHandler) scalingExecutor.getRejectedExecutionHandler()).rejected(),
+                rejectAfterShutdown ? equalTo((long) queuedAfterShutdown) : equalTo(0L)
+            );
+
+        } finally {
+            if (scalingExecutor.isShutdown() == false) {
```

Comment:
Should this not check against `isTerminated`? Perhaps just remove the `if`?

Reply:
I removed this `if` in ad729c90cdf439abdd2e713086bd947376988df9